lib/instrumentation/modules/@aws-sdk/client-sqs.js

```diff
@@ -0,0 +1,332 @@
+// DeleteMessageCommand => span.action = 'DeleteMessage'
+// do we need to have same messages (keep this key/val pairs) or its okay to have the ones form smithy-client?
+const OPERATIONS_TO_ACTIONS = {
+  DeleteMessage: 'delete',
+  DeleteMessageBatch: 'delete_batch',
+  ReceiveMessage: 'poll',
+  SendMessageBatch: 'send_batch',
+  SendMessage: 'send',
+  unknown: 'unknown',
+};
+const OPERATIONS = Object.keys(OPERATIONS_TO_ACTIONS);
+const MAX_SQS_MESSAGE_ATTRIBUTES = 10;
+const queueMetrics = new Map();
+
+/**
+ * Returns middlewares to instrument an S3Client instance
+ *
+ * @param {import('@aws-sdk/client-sqs').SQSClient} client
+ * @param {any} agent
+ * @returns {import('../@smithy/smithy-client').AWSMiddlewareEntry[]}
+ */
+function sqsMiddlewareFactory(client, agent) {
+  return [
+    {
+      middleware: (next, context) => async (args) => {
+        const ins = agent._instrumentation;
+        const log = agent.logger;
+        const span = ins.currSpan();
+        const input = args.input;
+
+        // W3C trace-context propagation.
+        const commandName = context.commandName.replace('Command', '');
+        const runContext = ins.currRunContext();
+        const parentSpan =
+          span || runContext.currSpan() || runContext.currTransaction();
+
+        if (parentSpan) {
+          const toPropagate = [];
+
+          if (commandName === 'SendMessage' && input.MessageAttributes) {
+            toPropagate.push(input.MessageAttributes);
+          } else if (
+            commandName === 'SendMessageBatch' &&
+            Array.isArray(input.Entries)
+          ) {
+            toPropagate.push(...input.Entries.map((e) => e.MessageAttributes));
+          }
+
+          // Though our spec only mentions a 10-message-attribute limit for *SQS*, we'll
+          // do the same limit here, because
+          // https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html
+          // mentions the 10-message-attribute limit for SQS subscriptions.
+          toPropagate.forEach((msgAttrs) => {
+            const attrsCount = Object.keys(msgAttrs).length + 2;
+            if (attrsCount > MAX_SQS_MESSAGE_ATTRIBUTES) {
+              log.warn(
+                { QueueUrl: input.QueueUrl },
+                'cannot propagate trace-context with SQS message, too many MessageAttributes',
+              );
+              return;
+            }
+            parentSpan.propagateTraceContextHeaders(
+              msgAttrs,
+              function (msgAttrs, name, value) {
+                if (name.startsWith('elastic-')) {
+                  return;
+                }
+                msgAttrs[name] = { DataType: 'String', StringValue: value };
+              },
+            );
+          });
+        }
+
+        // Ensure there is a span from the wrapped `client.send()`.
+        if (!span || !(span.type === TYPE && span.subtype === SUBTYPE)) {
+          return await next(args);
+        }
```

Comment:
Should this block be up at the top, as it is in client-s3.js and client-dynamodb.js?

I have the same question for client-sns.js.

Reply:
Checking the v2 implementations of both the trace context propagation is done always 1st and then there is the check of the `span` to decide if we have to extract info and add it to its context. I guess we want the same behaviour for v3

Reply:
You are correct. Thanks.